01/implement-pascals.checkpoint.md

```diff
@@ -0,0 +1,202 @@
+    and each index of the list stores the numbers in row `index + 1` 
+    of Pascal's triangle.
+    """
+    pass
+```
+### !end-placeholder
+### !tests
+```python
+import unittest
+from main import *
+
+class TestChallenge(unittest.TestCase):
+    def test_pascals_nominal(self):
+        # Arrange
+        numRows = 3
+
+        # Act
+        result = pascals_triangle(numRows)
+
+        # Assert
+        self.assertEqual(result, [[1], [1, 1], [1, 2, 1]])
+
+    def test_pascals_big_number(self):
+        # Arrange
+        numRows = 10
+
+        # Act
+        result = pascals_triangle(numRows)
+
+        # Act
+        self.assertEqual(result, [[1], [1, 1], [1, 2, 1], [1, 3, 3, 1], [1, 4, 6, 4, 1], [1, 5, 10, 10, 5, 1], [1, 6, 15, 20, 15, 6, 1], [1, 7, 21, 35, 35, 21, 7, 1], [1, 8, 28, 56, 70, 56, 28, 8, 1], [1, 9, 36, 84, 126, 126, 84, 36, 9, 1]])
+
+    def test_pascals_base_case(self):
+        # Arrange
+        numRows = 2
+
+        # Act
+        result = pascals_triangle(numRows)
+
+        # Assert
+        self.assertEqual(result, [[1], [1, 1]])
+
+    def test_pascals_out_of_range(self):
+        # Arrange
+        numRows = -1
+
+        # Assert
+        with self.assertRaises(ValueError):
+            pascals_triangle(numRows)
+```
+### !end-tests
+### !explanation
+
+An example of a working implementation:
+
+```python
+def pascals_triangle(num_rows):
+    if num_rows < 0:
+        raise ValueError("No input less than 0.")
+        
+    triangle = []
+
+    for row_num in range(num_rows):
+        # The first and last row elements are always 1.
+        row = [None for _ in range(row_num + 1)]
+        row[0], row[-1] = 1, 1
+
+        # Each triangle element is equal to the sum of the elements
+        # above-and-to-the-left and above-and-to-the-right.
+        for j in range(1, len(row) - 1):
+            row[j] = triangle[row_num - 1][j - 1] + triangle[row_num - 1][j]
+
+        triangle.append(row)
+
+    return triangle
+```
+### !end-explanation
+
+### !end-challenge
+<!-- prettier-ignore-end -->
+
+<!-- >>>>>>>>>>>>>>>>>>>>>> BEGIN CHALLENGE >>>>>>>>>>>>>>>>>>>>>> -->
+<!-- Replace everything in square brackets [] and remove brackets  -->
+
+### !challenge
+
+* type: paragraph
+* id: 2cf7eca9-1f15-47a3-a217-08dfdebbb71f
+* title: Time Complexity of Solution
+* points: 1
+
+##### !question
+
+What is the time complexity of your solution? Remember to define your valuables.
+
+##### !end-question
+
+##### !placeholder
+
+##### !end-placeholder
+
+### !end-challenge
+
+<!-- ======================= END CHALLENGE ======================= -->
+
+<!-- >>>>>>>>>>>>>>>>>>>>>> BEGIN CHALLENGE >>>>>>>>>>>>>>>>>>>>>> -->
+<!-- Replace everything in square brackets [] and remove brackets  -->
+
+### !challenge
+
+* type: paragraph
+* id: c48781b8-caf9-460a-af73-bf6e0b54585c
+* title: Space Complexity of Solution
+* points: 1 
+
+##### !question
+
+What is the space complexity of your solution? Remember to define your valuables.
```

Comment:
```suggestion
What is the space complexity of your solution? Remember to define your variables.
```

Reply:
Same as above. Instead of saying 'Remember to define your variables.' perhaps change to 'Please define and explain your variables.' Last cohort a lot of folks would define variables but not explain how they got to that time/space complexity